File: modules/vescape-core/ios/api/VescapeApi.swift
import Foundation

/// Authenticated HTTP boundary for every Vescape backend call made from native. Feature clients
/// describe a call; this owns the credential, the headers, the retry rule and the 401 policy, and
/// never learns what the call means.
///
/// It deliberately holds no Expo or React types: long-lived native work (recording, background
/// drains) must keep talking to the server while the JS runtime is gone.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/api/VescapeApi.kt
final class VescapeApi {
  private let baseUrl: String
  private let appVersion: String
  private let credentialProvider: () throws -> DeviceCredential?
  private let onUnauthorized: () throws -> Void
  private let transport: ApiTransport
  private let retryDelayNanoseconds: UInt64

  private static let invalidRequest = "invalid-request"
  private static let jsonContentType = "application/json"
  private static let retryDelayNanosecondsDefault: UInt64 = 300_000_000

  init(
    baseUrl: String,
    appVersion: String,
    credentialProvider: @escaping () throws -> DeviceCredential? = { nil },
    onUnauthorized: @escaping () throws -> Void = {},
    transport: ApiTransport = UrlSessionApiTransport(),
    retryDelayNanoseconds: UInt64 = VescapeApi.retryDelayNanosecondsDefault
  ) {
    self.baseUrl = baseUrl.hasSuffix("/") ? String(baseUrl.dropLast()) : baseUrl
    self.appVersion = appVersion
    self.credentialProvider = credentialProvider
    self.onUnauthorized = onUnauthorized
    self.transport = transport
    self.retryDelayNanoseconds = retryDelayNanoseconds
  }

  func request<T>(
    _ method: HttpMethod,
    path: String,
    query: [String: String] = [:],
    body: [String: Any]? = nil,
    auth: AuthMode = .required,
    parse: (String) throws -> T
  ) async -> ApiResult<T> {
    let token: String
    do { guard let resolved = try self.token(for: auth) else { return .unauthorized }; token = resolved }
    catch {
      if Self.isCancellation(error) { return .cancelled }
      UnexpectedNativeError.report(operation: "device_credential_read", category: "secure_store_read", error: error)
      return .unavailable("Device credential is unavailable")
    }
    let encodedBody = body.flatMap { serialize($0) }
    if body != nil && encodedBody == nil { return .malformed("Request body is not JSON") }
    let request = ApiRequest(
      method: method,
      url: url(path: path, query: query),
      headers: headers(token: token.isEmpty ? nil : token, hasBody: encodedBody != nil),
      body: encodedBody
    )
    return await send(
      request,
      rejectsStoredCredential: auth.rejectsStoredCredential && !token.isEmpty,
      parse: parse
    )
  }

  /// Resolved bearer token, empty when the call goes out anonymously, `nil` when a required
  /// credential is missing. A credential minted against another origin belongs to another
  /// environment, so it counts as missing rather than being sent to this one.
  private func token(for auth: AuthMode) throws -> String? {
    switch auth {
    case .bearer(let token): return token
    case .optional: return try storedToken() ?? ""
    case .required: return try storedToken()
    }
  }

  private func storedToken() throws -> String? {
    guard let credential = try credentialProvider() else { return nil }
    let origin = credential.serverUrl.hasSuffix("/")
      ? String(credential.serverUrl.dropLast())
      : credential.serverUrl
    return origin == baseUrl ? credential.token : nil
  }

  private func headers(token: String?, hasBody: Bool) -> [String: String] {
    var headers = [AppStatusCoordinator.appVersionHeader: appVersion]
    if hasBody { headers["Content-Type"] = Self.jsonContentType }
    if let token { headers["Authorization"] = "Bearer \(token)" }
    return headers
  }

  private func url(path: String, query: [String: String]) -> String {
    guard !query.isEmpty else { return baseUrl + path }
    var components = URLComponents()
    components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
    return baseUrl + path + "?" + (components.percentEncodedQuery ?? "")
  }

  private func send<T>(
    _ request: ApiRequest,
    rejectsStoredCredential: Bool,
    parse: (String) throws -> T
  ) async -> ApiResult<T> {
    var retried = false
    while true {
      do {
        let response = try await transport.execute(request)
        if response.status >= 500 {
          if canRetry(request, retried: retried) {
            retried = true
            do { try await Task.sleep(nanoseconds: retryDelayNanoseconds) }
            catch where Self.isCancellation(error) { return .cancelled }
            catch { return .unavailable(error.localizedDescription) }
            continue
          }
          return .unavailable("Server error (\(response.status))")
        }
        return outcome(response, rejectsStoredCredential: rejectsStoredCredential, parse: parse)
      } catch {
        if Self.isCancellation(error) { return .cancelled }
        if canRetry(request, retried: retried) {
          retried = true
          do { try await Task.sleep(nanoseconds: retryDelayNanoseconds) }
          catch where Self.isCancellation(error) { return .cancelled }
          catch { return .unavailable(error.localizedDescription) }
          continue
        }
        return .unavailable(error.localizedDescription)
      }
    }
  }

  /// Only idempotent calls repeat. No endpoint here takes a create key, so a repeated POST
  /// duplicates.
  private func canRetry(_ request: ApiRequest, retried: Bool) -> Bool {
    request.method.idempotent && !retried
  }

  private func outcome<T>(
    _ response: ApiResponse,
    rejectsStoredCredential: Bool,
    parse: (String) throws -> T
  ) -> ApiResult<T> {
    switch response.status {
    case 401:
      // An anonymous read cannot say anything about the stored credential, so it must not reject it.
      if rejectsStoredCredential {
        do { try onUnauthorized() }
        catch {
          if Self.isCancellation(error) { return .cancelled }
          UnexpectedNativeError.report(operation: "device_credential_reject", category: "secure_store_delete", error: error)
          return .unavailable("Device credential could not be rejected")
        }
      }
      return .unauthorized
    case 403: return .forbidden
    case 404: return .notFound
    case 400..<500: return .invalid(errorSlug(response.body))
    default:
      do {
        return .ok(try parse(response.body))
      } catch {
        if Self.isCancellation(error) { return .cancelled }
        return .malformed(error.localizedDescription)
      }
    }
  }

  private func errorSlug(_ body: String) -> String {
    guard let data = body.data(using: .utf8),
          // intentional-suppression: caller maps malformed request encoding to an explicit API error
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let error = json["error"] as? String,
          !error.isEmpty
    else { return Self.invalidRequest }
    return error
  }

  private func serialize(_ body: [String: Any]) -> String? {
    // intentional-suppression: caller maps malformed request encoding to an explicit API error
    guard let data = try? JSONSerialization.data(withJSONObject: body) else { return nil }
    return String(data: data, encoding: .utf8)
  }

  private static func isCancellation(_ error: Error) -> Bool {
    if Task.isCancelled || error is CancellationError { return true }
    return (error as? URLError)?.code == .cancelled
  }

  /// Client pinned to an origin. Callers pass the baked backend origin
  /// (`AppStatusCoordinator.serverBaseUrl`); provisioning passes the origin named by the credential
  /// it is about to verify, because nothing is stored yet. Reads stay usable while signed out — the
  /// credential is attached only when one exists.
  static func forOrigin(_ serverUrl: String) -> VescapeApi {
    let store = DeviceCredentialStore.shared
    return VescapeApi(
      baseUrl: serverUrl,
      appVersion: AppStatusCoordinator.installedMarketingVersion(),
      credentialProvider: { try store.read() },
      onUnauthorized: {
        try store.reject()
        Task { @MainActor in AppStatusCoordinator.shared.refresh() }
      }
    )
  }
}

/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/api/VescapeApi.kt `OkHttpApiTransport`
struct UrlSessionApiTransport: ApiTransport {
  private let timeout: TimeInterval = 10

  func execute(_ request: ApiRequest) async throws -> ApiResponse {
    guard let url = URL(string: request.url) else {
      throw NSError(domain: "VescapeApi", code: -1)
    }
    var urlRequest = URLRequest(url: url)
    urlRequest.httpMethod = request.method.rawValue
    urlRequest.timeoutInterval = timeout
    request.headers.forEach { urlRequest.setValue($0.value, forHTTPHeaderField: $0.key) }
    urlRequest.httpBody = request.body?.data(using: .utf8)
    let (data, response) = try await URLSession.shared.data(for: urlRequest)
    guard let http = response as? HTTPURLResponse else {
      throw NSError(domain: "VescapeApi", code: -2)
    }
    return ApiResponse(status: http.statusCode, body: String(data: data, encoding: .utf8) ?? "")
  }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/api/VescapeApi.kt
package expo.modules.vescapecore.api

import android.content.Context
import expo.modules.vescapecore.appstatus.AppStatusCoordinator
import expo.modules.vescapecore.auth.DeviceCredential
import expo.modules.vescapecore.auth.DeviceCredentialStore
import expo.modules.vescapecore.diagnostics.UnexpectedNativeError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * Authenticated HTTP boundary for every Vescape backend call made from native. Feature clients
 * describe a call; this owns the credential, the headers, the retry rule and the 401 policy, and
 * never learns what the call means.
 *
 * It deliberately holds no Expo or React types: long-lived native work (recording, background
 * drains) must keep talking to the server while the JS runtime is gone.
 *
 * @parity /modules/vescape-core/ios/api/VescapeApi.swift
 */
class VescapeApi(
  baseUrl: String,
  private val appVersion: String,
  private val credentialProvider: () -> DeviceCredential? = { null },
  private val onUnauthorized: () -> Unit = {},
  private val transport: ApiTransport = OkHttpApiTransport,
  private val retryDelayMillis: Long = RETRY_DELAY_MILLIS,
) {
  /** Trailing slash trimmed once, so path joins stay single-slashed and origin compares hold. */
  private val baseUrl = baseUrl.trimEnd('/')

  suspend fun <T> request(
    method: HttpMethod,
    path: String,
    query: Map<String, String> = emptyMap(),
    body: JSONObject? = null,
    auth: AuthMode = AuthMode.Required,
    parse: (String) -> T,
  ): ApiResult<T> = withContext(Dispatchers.IO) {
    val token = try {
      token(auth) ?: return@withContext ApiResult.Unauthorized
    } catch (error: Exception) {
      currentCoroutineContext().ensureActive()
      if (error is CancellationException) throw error
      UnexpectedNativeError.report("device_credential_read", "secure_store_read", error)
      return@withContext ApiResult.Unavailable("Device credential is unavailable")
    }
    val request = ApiRequest(
      method = method,
      url = url(path, query),
      headers = headers(token.ifEmpty { null }, body != null),
      body = body?.toString(),
    )
    send(request, rejectsStoredCredential = auth == AuthMode.Required && token.isNotEmpty(), parse = parse)
  }

  /**
   * Resolved bearer token, empty when the call goes out anonymously, `null` when a required
   * credential is missing. A credential minted against another origin belongs to another
   * environment, so it counts as missing rather than being sent to this one.
   */
  private fun token(auth: AuthMode): String? = when (auth) {
    is AuthMode.Bearer -> auth.token
    AuthMode.Optional -> storedToken() ?: ""
    AuthMode.Required -> storedToken()
  }

  private fun storedToken(): String? = credentialProvider()
    ?.takeIf { it.serverUrl.trimEnd('/') == baseUrl }
    ?.token

  private fun headers(token: String?, hasBody: Boolean): Map<String, String> = buildMap {
    put(AppStatusCoordinator.APP_VERSION_HEADER, appVersion)
    if (hasBody) put("Content-Type", JSON_CONTENT_TYPE)
    if (token != null) put("Authorization", "Bearer $token")
  }

  private fun url(path: String, query: Map<String, String>): String {
    val suffix = query.entries.joinToString("&") { (key, value) ->
      "${encode(key)}=${encode(value)}"
    }
    return baseUrl + path + if (suffix.isEmpty()) "" else "?$suffix"
  }

  private suspend fun <T> send(
    request: ApiRequest,
    rejectsStoredCredential: Boolean,
    parse: (String) -> T,
  ): ApiResult<T> {
    var retried = false
    while (true) {
      val response = try {
        transport.execute(request)
      } catch (e: Exception) {
        currentCoroutineContext().ensureActive()
        if (e is CancellationException) throw e
        if (canRetry(request, retried)) {
          retried = true
          delay(retryDelayMillis)
          continue
        }
        return ApiResult.Unavailable(e.message ?: e.javaClass.simpleName)
      }
      currentCoroutineContext().ensureActive()

      if (response.status >= 500) {
        if (canRetry(request, retried)) {
          retried = true
          delay(retryDelayMillis)
          continue
        }
        return ApiResult.Unavailable("Server error (${response.status})")
      }

      return outcome(response, rejectsStoredCredential, parse)
    }
  }

  /** Only idempotent calls repeat. No endpoint here takes a create key, so a repeated POST duplicates. */
  private fun canRetry(request: ApiRequest, retried: Boolean) = request.method.idempotent && !retried

  private fun <T> outcome(
    response: ApiResponse,
    rejectsStoredCredential: Boolean,
    parse: (String) -> T,
  ): ApiResult<T> {
    return when {
    response.status == 401 -> {
      // An anonymous read cannot say anything about the stored credential, so it must not reject it.
      if (rejectsStoredCredential) {
        try { onUnauthorized() }
        catch (error: Exception) {
          if (error is CancellationException) throw error
          UnexpectedNativeError.report("device_credential_reject", "secure_store_delete", error)
          return ApiResult.Unavailable("Device credential could not be rejected")
        }
      }
      ApiResult.Unauthorized
    }
    response.status == 403 -> ApiResult.Forbidden
    response.status == 404 -> ApiResult.NotFound
    response.status >= 400 -> ApiResult.Invalid(errorSlug(response.body))
    else -> try {
      ApiResult.Ok(parse(response.body))
    } catch (e: Exception) {
      if (e is CancellationException) throw e
      ApiResult.Malformed(e.message ?: e.javaClass.simpleName)
    }
    }
  }

  private fun errorSlug(body: String): String = try {
    JSONObject(body).optString("error").ifEmpty { INVALID_REQUEST }
  } catch (_: Exception) {
    INVALID_REQUEST
  }

  private fun encode(value: String): String = URLEncoder.encode(value, Charsets.UTF_8.name())

  companion object {
    private const val INVALID_REQUEST = "invalid-request"
    private const val JSON_CONTENT_TYPE = "application/json"
    private const val RETRY_DELAY_MILLIS = 300L

    /**
     * Client pinned to an origin. Callers pass the baked backend origin
     * ([AppStatusCoordinator.serverBaseUrl]); provisioning passes the origin named by the credential
     * it is about to verify, because nothing is stored yet. Reads stay usable while signed out — the
     * credential is attached only when one exists.
     */
    fun forOrigin(context: Context, serverUrl: String): VescapeApi {
      val app = context.applicationContext
      val store = DeviceCredentialStore(app)
      return VescapeApi(
        baseUrl = serverUrl.trimEnd('/'),
        appVersion = AppStatusCoordinator.get(app).appVersion,
        credentialProvider = { store.read() },
        onUnauthorized = {
          store.reject()
          AppStatusCoordinator.get(app).refresh()
        },
      )
    }
  }
}

/**
 * @parity /modules/vescape-core/ios/api/VescapeApi.swift `UrlSessionApiTransport`
 */
object OkHttpApiTransport : ApiTransport {
  private val client = OkHttpClient.Builder().callTimeout(10, TimeUnit.SECONDS).build()
  private val jsonMediaType = "application/json".toMediaType()

  override fun execute(request: ApiRequest): ApiResponse {
    val builder = Request.Builder().url(request.url)
    request.headers.forEach { (name, value) -> builder.header(name, value) }
    val body = when {
      request.body != null -> request.body.toRequestBody(jsonMediaType)
      request.method == HttpMethod.GET || request.method == HttpMethod.DELETE -> null
      else -> "".toRequestBody(jsonMediaType)
    }
    builder.method(request.method.name, body)
    return client.newCall(builder.build()).execute().use { response ->
      ApiResponse(response.code, response.body?.string().orEmpty())
    }
  }
}
